src/include/duckdb/common/types/string_type.hpp

```diff
@@ -70,6 +71,41 @@ struct string_t {
 	    : string_t(value.c_str(), UnsafeNumericCast<uint32_t>(value.size())) {
 	}
 
+	// Allocate using an arena allocator
+	string_t(const char *data, const uint32_t len, ArenaAllocator &arena) {
+		value.inlined.length = len;
+		D_ASSERT(data || GetSize() == 0);
+		if (IsInlined()) {
+			// zero initialize the prefix first
+			// this makes sure that strings with length smaller than 4 still have an equal prefix
+			memset(value.inlined.inlined, 0, INLINE_BYTES);
+			if (GetSize() == 0) {
+				return;
+			}
+			// small string: inlined
+			memcpy(value.inlined.inlined, data, GetSize());
+		} else {
```

Comment:
Maybe we can combine this code with the other `string_t` constructor via a function (in the header).